design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,387 @@
+-- fails with duplicate column error on 'V0'.
+
+-- Note: the equivalent query would of failed before the any-key-name work:
+CREATE TABLE OUTPUT AS SELECT ROWKEY, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- fails with duplicate column error on 'ROWKEY'.
+```
+
+As you can see from above, the common pattern of selecting the group by key and the aggregate fails.
+To 'fix' the query the user must remove `V0` from the projection - which is counter intuitive, as
+the user wishes this column in the result.
+
+We propose that this query should work, without any modification, and without storing duplicate
+data.
+
+### Requiring the key column in a projection in persistent queries
+
+By flipping the semantics so that the key column is not _implicitly_ copied across in persistent
+queries we place the burden on the user to _explicitly_ specify it in the projection.
+
+Where as it is possible to have a transient query without the key column in the projection, as a
+transient query is simply returning tabular data and has no concept of a key column, persistent
+queries do have the concept of a key column, and changing that column is only allowed via an
+explicit `PARTITION BY` or `GROUP BY` statement.  Therefore users will _always_ have to specify the
+key column in the projection of persistent queries.
+
+For example, the following statements, which previously executed, will now fail:
+
+```sql
+CREATE TABLE OUTPUT AS SELECT V0, V1 FROM INPUT;
+-- fails as key column not in projection.
+
+CREATE TABLE OUTPUT AS SELECT COUNT(*) FROM INPUT GROUP BY V1;
+-- fails as key column not in projection.
+
+CREATE STREAM OUTPUT AS SELECT V0, ID FROM INPUT PARTITION BY V1;
+-- fails as key column not in projection.
+
+CREATE TABLE OUTPUT AS SELECT I1.V0, I2.V2 FROM INPUT I1 JOIN INPUT I2 ON I1.ID = I2.ID;
+-- fails as key column not in projection.
+```
+
+We propose this is acceptable, and preferable to the current model.
+
+There is no reason why ksqlDB should not support creating sources in the future that do not contain
+the key column. See the [rejected alternatives](#rejected_alternatives) section for more info.
+
+### Allowing the key column to be copied to the value
+
+Hopefully, all of the above seems simple and clear. Now for one of the more tricky and murky bits...
+
+The observant among you may be thinking 'ah, but what if I _want_ to put a copy of the key column
+into the value?'. With current syntax you can do:
+
+```sql
+-- current syntax:
+CREATE TABLE OUTPUT AS SELECT V0, V1, ID AS V2 FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
+```
+
+With the proposed syntax the `ID AS V2` would be treated an aliased key column, resulting in the
+schema 'V2 INT KEY, V0 INT, V1 INT'. So, how does the user create an copy of the key in the value,
+if that's what they need to do?
+
+We propose introducing a `COPY` function, (name TBD - maybe 'AS_VALUE'?), that can be used to
```

Comment:
+1 for `AS_VALUE`, IMO `COPY` is too ambiguous

Reply:
Why do we need to introduce the special `COPY` function? With the changes in this KLIP, the key definition is inherited from the source in the `FROM` clause (unless we have a join, groupby or partition). Are the semantics that however many times a column (that is the key of the source) appears in the projection, every time it will be interpreted as the key? An alternative, which would be a bit more SQL like (in my mind at least) is to indicate the column that would be the primary key. So, something like:

```
CREATE TABLE OUTPUT AS SELECT PRIMARY(ID) , V0, V1, ID AS V2 FROM INPUT;
```
Just an idea though, not hung up on it. `COPY` gets the job done as well.

Reply:
@big-andy-coates what would happen with the following query?

```sql
CREATE TABLE OUTPUT (ID INT PRIMARY KEY, V0 INT, V1 INT, V2 INT) AS
  SELECT ID, V0, V1, ID AS V2 FROM INPUT
```
would it work without the `COPY` function?

I think the correct way to handle your query is having no key at all (which may be rejected until #5125 is done) unless a key is explicitly specified in the schema, like in the query I gave.

Since partial schemas are already a thing in ksqlDB (#4566) a key could be set like this:
```sql
CREATE TABLE OUTPUT (ID PRIMARY KEY) AS SELECT ID, V0, V1, ID AS V2 FROM INPUT
```
which looks quite similar to the one @vpapavas proposed.
This also works with composite keys in the future:

```sql
CREATE TABLE OUTPUT (PRIMARY KEY (ID, V1)) AS SELECT ID, V0, V1, ID AS V2 FROM INPUT
```

what do you think?

Reply:
@vpapavas 

> Why do we need to introduce the special `COPY` function? 

To provide a way for the user to add a copy of the key column as a value column.

> With the changes in this KLIP, the key definition is inherited from the source in the `FROM` clause (unless we have a join, groupby or partition). 

Confused... the _current_ behaviour is to inherit the key columns. The KLIP proposes removing this inheritance!  Not sure why you seem to think the KLIP is proposing key column inheritance...?

> Are the semantics that however many times a column (that is the key of the source) appears in the projection, every time it will be interpreted as the key? 

We only support a single key column at the moment, so no.  The issue is about allowing the user to copy the key into the value.

> An alternative, which would be a bit more SQL like (in my mind at least) is to indicate the column that would be the primary key. So, something like:
> 
> ```
> CREATE TABLE OUTPUT AS SELECT PRIMARY(ID) , V0, V1, ID AS V2 FROM INPUT;
> ```
> 
> Just an idea though, not hung up on it. `COPY` gets the job done as well.

Agreed this would work. However, it would logically follow that you should always need `PRIMARY` around the key column, which is a bit ugly.  Also, what about streams that don't have primary keys?

IMHO, the nice thing about `COPY` is that its only needed if you want to copy a key column into the value: i.e. there's no extra syntax for doing normal/easy things, only for doing more uncommon things.

Reply:
> @big-andy-coates what would happen with the following query?
> 
> ```sql
> CREATE TABLE OUTPUT (ID INT PRIMARY KEY, V0 INT, V1 INT, V2 INT) AS
>   SELECT ID, V0, V1, ID AS V2 FROM INPUT
> ```
> 
> would it work without the `COPY` function?
> 
> I think the correct way to handle your query is having no key at all (which may be rejected until #5125 is done) unless a key is explicitly specified in the schema, like in the query I gave.
> 
> Since partial schemas are already a thing in ksqlDB (#4566) a key could be set like this:
> 
> ```sql
> CREATE TABLE OUTPUT (ID PRIMARY KEY) AS SELECT ID, V0, V1, ID AS V2 FROM INPUT
> ```
> 
> which looks quite similar to the one @vpapavas proposed.
> This also works with composite keys in the future:
> 
> ```sql
> CREATE TABLE OUTPUT (PRIMARY KEY (ID, V1)) AS SELECT ID, V0, V1, ID AS V2 FROM INPUT
> ```
> 
> what do you think?

Hey @PeterLindner, that's an interesting idea!

I'm not sure I agree the table should default to no primary key. That would mean more verbose syntax for the common case. (Plus, requiring KSQL to support key-less table MVs). Though, I guess its strictly more standard SQL as a RDBS wouldn't create OUTPUT with a primary key by default (well, I don't think it would - I'd need to check and sqlfiddle.com is down ;) ).

However, this may be a better syntax for handling the edgecase where the user wants to put a copy of the key column in the value, e.g. 

```sql
CREATE TABLE OUTPUT (ID PRIMARY KEY) AS SELECT ID, V0, V1, ID AS V2 FROM INPUT
-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT

CREATE STREAM OUTPUT (ID KEY) AS SELECT ID, V0, V1, ID AS V2 FROM INPUT
-- resulting schema: ID INT KEY, V0 INT, V1 INT, V2 INT
```

The main downside of this approach is that it's more syntax / more work.

Thoughts others? @derekjn @rmoff @agavra @mjsax etc?

Reply:
Just realized I forgot the type in the partial schema... instead of defining the type as well like:
```sql
CREATE TABLE OUTPUT (
ID INT PRIMARY KEY
-- V0 INT omitted
-- V1 INT omitted
-- V2 INT omitted
) ... 
```
The following partial schema may be more intuitive:
```sql
CREATE TABLE OUTPUT (
-- ID INT omitted
-- V0 INT omitted 
-- V1 INT omitted 
-- V2 INT omitted 
PRIMARY KEY (ID) 
) ... 
```
And analogous `KEY (ID)` for streams.

Furthermore I realized that people may try to repartition a table with this syntax like:
```sql
CREATE TABLE (PRIMARY KEY (V0)) AS SELECT ... 
-- expecting schema: V0 INT KEY, ID INT, V1 INT, V2 INT
```
don't know if that's a problem

Reply:
Interesting idea. I am not sure though if this would work? I think in SQL, a `CRATE TABLE output <schema> AS SELECT` requires that the result schema of the query it _compatible_ to the schema definition, and thus the schema definition does not modify the query result/schema (what make a lot of sense to me; for example if you specify a PK constraint and the query result would violate the PK constraint the query would fail and the table would not even be created). The query semantics must be defined by the query itself, and must be the same w/ or w/o a preceding `CT` clause.

To me, it seems that the proposal tried to put the cart before the horse.

Reply:
@mjsax this means the select would need a `PARTITION BY` to be compatible with the schema in the first place? That sounds logical...

Would it make sense to allow to `PARTITION BY` an already existing key for a table? That's either redundant or solves the edge case

Reply:
Atm, we consider it redundant that thus don't even allow `PARTITION BY` for tables (If one does a `GROUP BY a` it's implicitly partitioned by `a`). It's only allowed for streams as there, it won't really change the schema (there is no PK) but only change the "schema to kv mapping" and thus the partitioning.

Reply:
This is a great discussion and I like the syntax. However, for the scope of this KLIP I'm going to go with a slightly more hack solution for the short to mid term. Otherwise, this grows to large!  Really appreciate your input @PeterLindner, as always.

Reply:
@big-andy-coates I'm glad my random ideas are appreciated :smile:

I also found this [MySQL doc for CREATE TABLE ... SELECT](https://dev.mysql.com/doc/refman/8.0/en/create-table-select.html) which technically is not the same because it does not create a materialized view but may help if you decide to work on the syntax later